File: lib/models/properties/note.dart
import '../../utils/json_helpers.dart';
import 'property_metadata.dart';

/// Note property.
///
/// Android allows multiple notes per contact; iOS/macOS allows only one.
///
/// iOS requires entitlements and `FlutterContacts.config.enableIosNotes = true`.
/// See [Apple's documentation](https://developer.apple.com/documentation/bundleresources/entitlements/com.apple.developer.contacts.notes).
class Note {
  /// Note text.
  final String note;

  /// Property identity metadata (used internally for updates).
  final PropertyMetadata? metadata;

  const Note({required this.note, this.metadata});

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'note': note};
    JsonHelpers.encode(json, 'metadata', metadata, (m) => m.toJson());
    return json;
  }

  static Note fromJson(Map json) => Note(
    note: json['note'] as String,
    metadata: JsonHelpers.decode(json['metadata'], PropertyMetadata.fromJson),
  );

  @override
  String toString() => JsonHelpers.formatToString('Note', toJson());

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Note && note == other.note);

  @override
  int get hashCode => note.hashCode;

  Note copyWith({String? note, PropertyMetadata? metadata}) {
    return Note(note: note ?? this.note, metadata: metadata ?? this.metadata);
  }
}

File: macos/flutter_contacts/Sources/flutter_contacts/crud/models/properties/Note.swift
import Contacts
import Foundation

struct Note: Equatable, Hashable {
    let note: String
    let metadata: PropertyMetadata?

    init(note: String, metadata: PropertyMetadata? = nil) {
        self.note = note
        self.metadata = metadata
    }

    func toJson() -> Json {
        var json: Json = ["note": note]
        json.setJson("metadata", metadata) { $0.toJson() }
        return json
    }

    static func fromJson(_ json: Json) -> Note {
        Note(
            note: json["note"] as! String,
            metadata: json.json("metadata", PropertyMetadata.fromJson)
        )
    }

    static func apply(_ note: String?, to cnContact: CNMutableContact) {
        cnContact.note = note ?? ""
    }
}
